iceoryx_utils/platform/mac/source/time.cpp

```diff
@@ -0,0 +1,146 @@
+            ulock, [timerid] { return timerid->parameter.isTimerRunning || !timerid->keepRunning.load(); });
+    }
+
+    return timerid->parameter.isTimerRunning;
+}
+
+static void
+setTimeParameters(timer_t timerid, const itimerspec& timeParameters, const bool runOnce, const bool isTimerRunning)
+{
+    std::lock_guard<std::mutex> l(timerid->parameter.mutex);
+    clock_gettime(CLOCK_REALTIME, &timerid->parameter.startTime);
+    timerid->parameter.timeParameters = timeParameters;
+    timerid->parameter.runOnce = runOnce;
+    timerid->parameter.wasCallbackCalled = false;
+    timerid->parameter.isTimerRunning = isTimerRunning;
+}
+
+int timer_create(clockid_t clockid, struct sigevent* sevp, timer_t* timerid)
+{
+    timer_t timer = new appleTimer_t();
+    timer->callback = sevp->sigev_notify_function;
+    timer->callbackParameter = sevp->sigev_value;
+
+    timer->thread = std::thread([timer] {
+        while (timer->keepRunning.load())
+        {
+            if (waitForExecution(timer))
+            {
+                timer->parameter.mutex.lock();
+                bool doCallCallback =
+                    !timer->parameter.runOnce || timer->parameter.runOnce && !timer->parameter.wasCallbackCalled;
+                timer->parameter.mutex.unlock();
+                if (doCallCallback)
+                {
+                    timer->callback(timer->callbackParameter);
+
+                    std::lock_guard<std::mutex> l(timer->parameter.mutex);
+                    timer->parameter.wasCallbackCalled = true;
+                }
+            }
+        }
+    });
+
+    *timerid = timer;
+    return 0;
+}
+
+int timer_delete(timer_t timerid)
+{
+    stopTimerThread(timerid);
+    delete timerid;
+    return 0;
+}
+
+int timer_settime(timer_t timerid, int flags, const struct itimerspec* new_value, struct itimerspec* old_value)
+{
+    // disarm timer
+    if (new_value->it_value.tv_sec == 0 && new_value->it_value.tv_nsec == 0)
+    {
+        setTimeParameters(timerid, *new_value, false, false);
+    }
+    // run once
+    else if (new_value->it_interval.tv_sec == 0 && new_value->it_interval.tv_nsec == 0)
+    {
+        setTimeParameters(timerid, *new_value, true, true);
+    }
+    // run periodically
+    else
+    {
+        setTimeParameters(timerid, *new_value, false, true);
+    }
+    timerid->parameter.wakeup.notify_one();
+
+    return 0;
+}
+
+int timer_gettime(timer_t timerid, struct itimerspec* curr_value)
+{
+    timespec currentTime, startTime;
+    clock_gettime(CLOCK_REALTIME, &currentTime);
+    {
+        std::lock_guard<std::mutex> l(timerid->parameter.mutex);
+        curr_value->it_interval = timerid->parameter.timeParameters.it_interval;
+        startTime = timerid->parameter.startTime;
+    }
+    curr_value->it_value.tv_sec = curr_value->it_interval.tv_sec - (currentTime.tv_sec - startTime.tv_sec);
+    curr_value->it_value.tv_nsec = curr_value->it_interval.tv_nsec - (currentTime.tv_nsec - startTime.tv_nsec);
```

Comment:
the calculation could become negative, are positive seconds and negative nanoseconds allowed?